### PR TITLE
Feature/COR-1170-rioolwaterpagina-verbetering-choropletenkaart-nl

### PR DESCRIPTION
**Pull request by VWSCoronaDashboard28**
# Summary

This PR addresses the changes described in COR-1170. 
TLDR: This PR implements a solution that highlights areas within a choropleth with outdated data scoped to waste water.

**Changes:**
- Added a notification to the tooltip of a given area when its data is outdated
- Added a new item to the legend for the choropleth for waste water
- Added a new color which is used for an area with outdated data
- Added new `data_is_outdated` property to the schemas for GM and VR collections
- Added some new sanity keys for text

**NOTE:** I am unsure weather this PR should receive the label `Hold`. Maybe it should as the data is not added in the backend yet (AFAIK).

### BEFORE (Production)

**GM**
BEFORE

**VR**
BEFORE-VR

### AFTER (Local)

**GM**
AFTER

**VR**
AFTER-VR

### CODE REVIEW

File: packages/app/src/components/choropleth/tooltips/tooltip.tsx [hunk 1]
```diff
@@ -28,25 +28,14 @@ const padding = {
   top: 12,
```
Comment (VWSCoronaDashboard28) [inline packages/app/src/components/choropleth/tooltips/tooltip.tsx:28]: Changes in this file seem to be caused by formatting, you can skip this file while reviewing if you'd like to.

File: packages/app/src/pages/landelijk/rioolwater.tsx [hunk 2]
```diff
@@ -195,7 +156,9 @@ const SewerWater = (props: StaticProps<typeof getStaticProps>) => {
             legend={{
               title: textNl.legenda_titel,
               thresholds: thresholds.vr.average,
+              outdatedDataLabel: textNl.choropleth_legend_outdated_data_label,
```
Comment (VWSCoronaDashboard28) [inline packages/app/src/pages/landelijk/rioolwater.tsx:159]: Adding `outdatedDataLabel` here and `pageType` below are the only changes in this file, the rest seem to be formatting.

File: packages/cms/src/schemas/schema.ts [hunk 3]
```diff
@@ -24,6 +24,6 @@ export default createSchema({
     ...Object.values(objects),
```
Comment (VWSCoronaDashboard28) [inline packages/cms/src/schemas/schema.ts:24]: Did not touch this file

File: packages/cms/src/schemas/topical/measure-tile-collection.ts [hunk 4]
```diff
@@ -1,4 +1,3 @@
-
 import { REQUIRED } from '../../validation';
```
Comment (VWSCoronaDashboard28) [inline packages/cms/src/schemas/topical/measure-tile-collection.ts:1]: Did not touch this file.

File: packages/cms/src/schemas/topical/measure-tile.ts [hunk 5]
```diff
@@ -25,4 +25,4 @@ export const measureTile = {
       title: 'description.nl',
```
Comment (VWSCoronaDashboard28) [inline packages/cms/src/schemas/topical/measure-tile.ts:25]: Did not touch this file

File: packages/cms/src/schemas/topical/thermometer.ts [hunk 6]
```diff
@@ -19,12 +19,12 @@ export const thermometer = {
     {
```
Comment (VWSCoronaDashboard28) [inline packages/cms/src/schemas/topical/thermometer.ts:19]: Did not touch this file

File: packages/cms/src/validation/index.ts [hunk 7]
```diff
@@ -1,5 +1,4 @@
-
 import { Rule } from '~/sanity';
```
Comment (VWSCoronaDashboard28) [inline packages/cms/src/validation/index.ts:1]: Did not touch this file

File: packages/app/src/components/choropleth/tooltips/choropleth-tooltip.tsx [hunk 8]
```diff
@@ -4,71 +4,51 @@ import { Markdown, VisuallyHidden } from '~/components';
...
-    `[${
-      ChoroplethTooltip.name
-    }] No tooltip content found in siteText.choropleth_tooltip.${
-      data.map
-    }.${data.dataConfig.metricProperty.toString()}`
-  );
+  const tooltipContent = (text as unknown as Record<string, Record<string, Record<string, string>>>)[data.map]?.[data.dataConfig.metricProperty as string]?.content;
+  assert(isDefined(tooltipContent), `[${ChoroplethTooltip.name}] No tooltip content found in siteText.choropleth_tooltip.${data.map}.${data.dataConfig.metricProperty.toString()}`);
+
+  let tooltipNotification;
```
Comment (VWSCoronaDashboard19) [inline packages/app/src/components/choropleth/tooltips/choropleth-tooltip.tsx:31]: Is it possible to move this to a `useState` hook? same for line 68-69.

File: packages/app/src/components/choropleth-legenda.tsx [hunk 9]
```diff
@@ -96,23 +90,29 @@ const List = styled.ul(
...
 );
 
-const Item = styled.li(
+// Assigning a flex-basis based on number of items in threshold so that legend items outside of the threshold array can be displayed using the same measurement.
+const Item = styled.li<{
+  width: number;
+}>(({ width }) =>
   css({
     flex: 1,
+    flexBasis: `${width}%`,
```
[the remark this answers is not in the document]
Comment (VWSCoronaDashboard28) [reply packages/app/src/components/choropleth-legenda.tsx:99]: For this one on the `Item` component I'm not passing the width as a percentage in the prop therefore you see it like this, but for the `LegendaColor`, I am. I can change it to also pass the percentage directly for the `Item` component if you prefer that.

File: packages/app/src/components/choropleth/tooltips/choropleth-tooltip.tsx [hunk 10]
```diff
@@ -85,25 +65,29 @@ export function ChoroplethTooltip<T extends ChoroplethDataItem>(
...
+      <TooltipSubject subject={replaceVariablesInText(subject, tooltipVars)} thresholdValues={data.thresholdValues} filterBelow={filterBelow}>
         <VisuallyHidden>{ariaContent}</VisuallyHidden>
         <Box aria-hidden>
           <Markdown content={content} />
         </Box>
       </TooltipSubject>
+
+      {showNotification && outdatedDataDate && tooltipNotification && (
+        <TooltipNotification>
+          <Markdown content={replaceVariablesInText(tooltipNotification, { date: outdatedDataDate })} />
```
Comment (VWSCoronaDashboard19) [inline packages/app/src/components/choropleth/tooltips/choropleth-tooltip.tsx:88]: Does this also read out the yellow notification text on GMs and VRs with missing data, when using a screenreader on the choropleth?

File: packages/app/src/components/choropleth-legenda.tsx (hunk 9 above)
[the remark this answers is not in the document]
Comment (VWSCoronaDashboard19) [reply packages/app/src/components/choropleth-legenda.tsx:99]: Ah I did not see the `%`, apologies. Think it's good for now.

File: packages/app/src/components/choropleth/tooltips/choropleth-tooltip.tsx (hunk 8 above)
Comment (VWSCoronaDashboard28) [reply packages/app/src/components/choropleth/tooltips/choropleth-tooltip.tsx:31]: Yes it's possible, however I'm wondering whether it will make much of a difference. I know that it is favourable to use `useState` as it will keep the `view` up to date with any variable changes. However the notification text, whether to show the notification, and the outdated date is not really subject to change unless new data is available.

Now that you mention it, state might be a better way to go, but I'm not sure. 🤔

File: packages/app/src/components/choropleth/tooltips/choropleth-tooltip.tsx (hunk 10 above)
Comment (VWSCoronaDashboard28) [reply packages/app/src/components/choropleth/tooltips/choropleth-tooltip.tsx:88]: It doesn't, shall I add a `VisuallyHidden` component to it the same way we do it on line 80, passing it the `aria-content` for the notification?

File: packages/app/src/components/choropleth/tooltips/tooltip-subject.tsx [hunk 11]
```diff
@@ -29,30 +24,30 @@ export function TooltipSubject({
...
+    </StyledTooltipSubject>
   );
 }
+
+const StyledTooltipSubject = styled.div(
+  css({
+    py: 2,
+    px: 3,
+  })
+);
```
Comment (VWSCoronaDashboard26) [inline packages/app/src/components/choropleth/tooltips/tooltip-subject.tsx:53]: As per earlier discussions, this should be written as `paddingY: space[2]` and `paddingX: space[3]`. You could also opt to completely take out using the `css()` method and shorthand padding.
```
const StyledTooltipSubject = styled.div`
  padding: ${space[2]} ${space[3]};
`;
```

File: packages/app/src/components/choropleth/tooltips/choropleth-tooltip.tsx (hunk 10 above)
Comment (VWSCoronaDashboard19) [reply packages/app/src/components/choropleth/tooltips/choropleth-tooltip.tsx:88]: Yes, would be nice if these specific cases would also be mentioned by screenreaders.

File: packages/app/src/components/choropleth/tooltips/tooltip-subject.tsx [hunk 12]
```diff
@@ -29,30 +24,30 @@ export function TooltipSubject({
...
+          spacingHorizontal={2}
+          css={css({
+            display: 'flex',
+            alignItems: 'center',
+            flexWrap: 'nowrap',
+            whiteSpace: 'pre-wrap',
+          })}
+        >
+          {children}
+          <Box flexShrink={0} height={13} width={13} borderRadius={'2px'} ml={'auto'} mb={'auto'} mt={'5px'} backgroundColor={color} />
```
Comment (VWSCoronaDashboard26) [inline packages/app/src/components/choropleth/tooltips/tooltip-subject.tsx:41]: I assume that the values for `height`, `width` and such are pixel values as they are outside of the `space` array range. If that is the case, I would propose to write these sort of props as `height="13px"`, just for added clarity.

**Review by VWSCoronaDashboard26 — COMMENTED** [review]
Decided to leave this as a generic comment as I do not want to clutter this pull request with repetitive comments, but have a look at files with abbreviated props such as `pr`, `m`, etc.

File: packages/app/src/components/choropleth/tooltips/choropleth-tooltip.tsx (hunk 8 above)
Comment (VWSCoronaDashboard28) [reply packages/app/src/components/choropleth/tooltips/choropleth-tooltip.tsx:31]: I am trying this out right now, first with the tooltip notification, and I get a `Too many re-renders` error. I can avoid that by setting the state within a `useEffect` but then then the `useEffect` would run also for non-sewer map choropleths. 

So i'm not sure whats the best way to proceed 🤔

File: packages/app/src/components/choropleth/tooltips/choropleth-tooltip.tsx (hunk 10 above)
Comment (VWSCoronaDashboard28) [reply packages/app/src/components/choropleth/tooltips/choropleth-tooltip.tsx:88]: Yup, I agree, ill do this, and thanks for the pointer!

File: packages/app/src/components/choropleth/tooltips/choropleth-tooltip.tsx [hunk 13]
```diff
@@ -4,71 +4,51 @@ import { Markdown, VisuallyHidden } from '~/components';
...
-    formatNumber,
-    formatPercentage,
-    formatDate,
-    formatDateFromSeconds,
-    formatDateFromMilliseconds,
-    formatRelativeDate,
-    formatDateSpan,
-  } = useIntl();
+  const { commonTexts, formatNumber, formatPercentage, formatDate, formatDateFromSeconds, formatDateFromMilliseconds, formatRelativeDate, formatDateSpan } = useIntl();
+  const isSewerMap = data.dataConfig.metricName === 'sewer';
```
[the remark this answers is not in the document]
Comment (VWSCoronaDashboard28) [reply packages/app/src/components/choropleth/tooltips/choropleth-tooltip.tsx:21]: You mean something like `tooltipVars.data_is_outdated` on line 71 and then scope everything to that condition?

File: packages/app/src/components/choropleth-legenda.tsx [hunk 14]
```diff
@@ -44,13 +42,18 @@ export function ChoroplethLegenda({
...
-    <Box
-      width="100%"
-      pr={`${endLabelSize.width ?? 0 / 2}px`}
-      spacing={2}
-      aria-hidden="true"
-    >
+    <Box width="100%" pr={`${endLabelSize.width ?? 0 / 2}px`} spacing={2} aria-hidden="true">
       {title && <Text variant="subtitle1">{title}</Text>}
+
+      {pageType === 'sewer' && (
```
[the remark this answers is not in the document]
Comment (VWSCoronaDashboard28) [reply packages/app/src/components/choropleth-legenda.tsx:48]: I think I do agree with the logic of what you're saying here. But at this point, it would be an array with a single entry so I'm not sure how much value it will add at the moment.

I do think that it should be done the way you are suggesting, but maybe at a point where we know that other choropleths might need the same logic. What do you think?

File: packages/app/src/components/choropleth-legenda.tsx [hunk 15]
```diff
@@ -60,20 +63,11 @@ export function ChoroplethLegenda({
             const formattedTreshold = formatNumber(threshold);
 
             return (
-              <Item
-                key={color + threshold}
-                ref={index === 0 ? itemRef : undefined}
-              >
+              <Item key={color + threshold} ref={index === 0 ? itemRef : undefined} width={100 / thresholds.length}>
```
[the remark this answers is not in the document]
Comment (VWSCoronaDashboard28) [reply packages/app/src/components/choropleth-legenda.tsx:66]: Hasan had a similar comment, I've changed it so that the width prop passes in the percentage string in both cases.

File: packages/app/src/components/choropleth/tooltips/tooltip-notification.tsx [hunk 16]
```diff
@@ -0,0 +1,17 @@
+import { Box } from '~/components/base';
+import { colors } from '@corona-dashboard/common';
+import { space } from '~/style/theme';
+
+interface TooltipNotificationProps {
+  children: React.ReactNode;
+}
+
+export const TooltipNotification = (props: TooltipNotificationProps): React.ReactElement => {
```
[the remark this answers is not in the document]
Comment (VWSCoronaDashboard28) [reply packages/app/src/components/choropleth/tooltips/tooltip-notification.tsx:9]: I guess I probably don't, still figuring out where and when to use Typescript like this 😄

File: packages/app/src/components/choropleth/tooltips/tooltip-subject.tsx (hunk 12 above)
Comment (VWSCoronaDashboard28) [reply packages/app/src/components/choropleth/tooltips/tooltip-subject.tsx:41]: I would assume so as well, however I did not change these values so I'm not sure. But i can make these suggested changes.

File: packages/app/src/components/choropleth/tooltips/choropleth-tooltip.tsx (hunk 10 above)
Comment (VWSCoronaDashboard28) [reply packages/app/src/components/choropleth/tooltips/choropleth-tooltip.tsx]: All done 😄

File: packages/app/src/components/choropleth-legenda.tsx (hunk 14 above)
[the remark this answers is not in the document]
Comment (VWSCoronaDashboard26) [reply packages/app/src/components/choropleth-legenda.tsx:48]: I agree. Like I said: I might have taken it a bit overboard. 😅 It is fine to leave this like this, but let's try to make some type of mental note of this comment when something similar might have to be developed.

File: packages/cms/src/lokalize/key-mutations.csv [hunk 17]
```diff
@@ -1,4 +1,109 @@
...
+2022-11-01T13:08:15.181Z,add,common.variant_codes.XBB,YyKAMdtTUUtpI3N1EArBHE,__
+2022-10-19T12:59:20.133Z,add,pages.behavior_page.nl.section_archived.description,ibZWLQHfYl2AyCZBOq5lUb,__
+2022-10-19T12:59:21.053Z,add,pages.behavior_page.nl.section_archived.title,rSoN2URF0rxEHQvXsrmWoN,__
+2022-10-20T12:37:16.287Z,add,common.behavior.subjects.posttest_isolation,ibZWLQHfYl2AyCZBOySw6K,__
+2022-10-20T08:52:25.474Z,delete,pages.intensive_care_page.nl.kpi_bedbezetting.barscale_screenreader_text,jDbv30Y7XYH5kl8ZaovZD6,__
+2022-11-07T15:09:11.939Z,add,common.choropleth_tooltip.gm.outdated_data_notification,TrUIlGMFRR6fBR627mmYKT,__
+2022-11-07T15:41:35.054Z,delete,common.choropleth_tooltip.gm.outdated_data_notification,TrUIlGMFRR6fBR627mmYKT,__
+2022-11-07T15:57:14.611Z,add,common.choropleth_tooltip.outdated_data_notification,Fgb5SAAswLDPRQ9KRzHRc8,__
+2022-11-07T16:00:34.118Z,delete,common.choropleth_tooltip.outdated_data_notification,Fgb5SAAswLDPRQ9KRzHRc8,__
+2022-11-07T16:01:10.105Z,add,common.choropleth_tooltip.gm.average.outdated_data_notification,Fgb5SAAswLDPRQ9KRzIAnG,__
```
Comment (VWSCoronaDashboard28) [inline packages/cms/src/lokalize/key-mutations.csv:104]: @VWSCoronaDashboard19 The keys on this line, 105 and 106, were the only ones added by me for this feature. However, when I merged develop into my branch I noticed a lot of changed in this file. @VWSCoronaDashboard26 Pointed out that it could be related to the fact that https://github.com/minvws/nl-covid19-data-dashboard/pull/4494 is not merged yet. Is that assumption accurate?

File: packages/cms/src/lokalize/key-mutations.csv [hunk 18]
```diff
@@ -1,4 +1,109 @@
...
+2022-11-07T16:01:11.103Z,add,common.choropleth_tooltip.vr.average.outdated_data_notification,BBCOmHCrMQzaQKV1wWKvZ9,__
+2022-11-11T08:49:48.826Z,add,pages.sewer_page.nl.choropleth_legend_outdated_data_label,TrUIlGMFRR6fBR628D5VMd,__
 2022-11-07T15:57:59.982Z,delete,pages.topical_page.nl.thermometer.article_reference,NYpb1u6P0tqViLcixo7iYI,__
```
Comment (VWSCoronaDashboard28) [inline packages/cms/src/lokalize/key-mutations.csv:107]: @VWSCoronaDashboard19 The keys on line 104, 105 and 106, were the only ones added by me for this feature. However, when I merged develop into my branch I noticed a lot of changed in this file. @VWSCoronaDashboard26 Pointed out that it could be related to the fact that https://github.com/minvws/nl-covid19-data-dashboard/pull/4494 is not merged yet. Is that assumption accurate?

Comment (VWSCoronaDashboard19) [reply packages/cms/src/lokalize/key-mutations.csv:107]: Try to merge it again as that PR is closed now, the deletions should be gone.

File: packages/cms/src/lokalize/key-mutations.csv [hunk 19]
```diff
@@ -1 +1,8 @@
 timestamp,action,key,document_id,move_to
```
Comment (VWSCoronaDashboard28) [inline packages/cms/src/lokalize/key-mutations.csv:1]: @VWSCoronaDashboard26 and @VWSCoronaDashboard19 - This looks better right? I merged in develop again, had some conflicts, so I took only the things I added in my branch 🤷 .

Comment (VWSCoronaDashboard19) [reply packages/cms/src/lokalize/key-mutations.csv:1]: Bellissimo 👌

**Review by VWSCoronaDashboard26 — APPROVED** [review]
Approved, but let's not merge this until all schemas can be validated.

**Comment by VWSCoronaDashboard28** [comment]
@VWSCoronaDashboard26 and @VWSCoronaDashboard19 I've pushed some new changes in the latest commit. Apparently backend needs the datapoint to be present for sewers on all VR and GM areas, so I've added it there as well.

**Comment by VWSCoronaDashboard19** [comment]
> @VWSCoronaDashboard26 and @VWSCoronaDashboard19 I've pushed some new changes in the latest commit. Apparently backend needs the datapoint to be present for sewers on all VR and GM areas, so I've added it there as well.

Can you also generate the corresponding data types for your schema changes?

**Comment by VWSCoronaDashboard28** [comment]
> > @VWSCoronaDashboard26 and @VWSCoronaDashboard19 I've pushed some new changes in the latest commit. Apparently backend needs the datapoint to be present for sewers on all VR and GM areas, so I've added it there as well.
> 
> Can you also generate the data types for your schema changes?

@VWSCoronaDashboard19 Done! However, I do believe that this command runs when deploying to any environment, so technically it's not needed right? But i do understand that it helps other devs when pulling in the code.

**Comment by VWSCoronaDashboard19** [comment]
> @VWSCoronaDashboard19 Done! However, I do believe that this command runs when deploying to any environment, so technically it's not needed right? But i do understand that it helps other devs when pulling in the code.

Not sure, but I believe it is a manual task to generate types. If types were incorrect, this branch was not mergable due to failing checks on this PR.

**Comment by VWSCoronaDashboard28** [comment]
> Not sure, but I believe it is a manual task to generate types. If types were incorrect, this branch was not mergable due to failing checks on this PR (or a failing pipeline).

Aha, thats good to know. Thanks!
